### ISSUE TITLE
Documentation for Testing package

### ISSUE THREAD

**Issue by RobertBroersma** [redwoodjs/redwoodjs.com#162]
Here's a small doc on how to (currently) work with the `@redwoodjs/testing` package.

@jtoar @thedavidprice I'm still a bit lost in the whole docs situation so I was hoping one of you could put this in the right place! Feel free to make any changes and add stuff as well—I'm not much of a writer 🤷‍♂️ 

---

# Testing
Redwood provides several helpers to make your life easier when testing your Redwood application. It is a simple wrapper around [@testing-library/react](https://testing-library.com/docs/react-testing-library/intro) with a coat of sugar to make it work out of the box with other Redwood packages like `@redwoodjs/router`. You can import anything you'd normally import from `@testing-library/react` from `@redwoodjs/testing`.

Here's an example:
```javascript
import { render, screen } from '@redwoodjs/testing'

import BlogLayout from './BlogLayout'

test('Loading renders successfully', () => {
  render(<BlogLayout />)

  expect(screen.getByText('Blog')).toBeInTheDocument()
})
```

# Running tests
You can run your test suite by running `yarn rw test` in your terminal.

Refer to the [command line interface](https://redwoodjs.com/docs/command-line-interface#test) for more info.

# Mocking API calls
Redwood uses Apollo Client under the hood. Apollo provides a `MockedProvider` that you can use to mock the result of calls your app is making to the API side. You'll

Here's an example:
```javascript
import { render, screen } from '@redwoodjs/testing'
import { MockedProvider } from '@apollo/react-testing'

import BlogPostsCell, { QUERY } from './BlogPostsCell'

test('Success renders successfully', async () => {
  const mocks = [
    {
      request: {
        query: QUERY,
        variables: {},
      },
      result: {
        data: {
          posts: [
            {
              id: 'id-1',
              title: 'Post 1',
            },
          ],
        },
      },
    },
  ]

  render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <BlogPostsCell />
    </MockedProvider>
  )

  expect(await screen.findByText(/Post 1/i)).toBeInTheDocument()
})
```

Check [Testing React Components](https://www.apollographql.com/docs/react/development-testing/testing/) in the Apollo docs for examples on testing Loading and Error states.

**Comment by jtoar** [comment]
@robert I think this is a great start, thanks a ton for doing this!

@thedavid I started a google doc for this: https://docs.google.com/document/d/1bUnj1UUUK29HrojgKCzXLCHdsalOVsKyn0l9YLnnY-0/edit?usp=sharing.

Robert, for a few other docs before this David and I used a google doc to write everything, an issue (like this one) to talk over changes/ask for reviews on our edits, then wrap it all up in a PR when it was done. We could try that here again? @thedavidprice did you think that workflow worked before?

**Comment by RobertBroersma** [comment]
Sounds like a plan!

**Comment by thedavidprice** [comment]
@jtoar Google docs process seems fine as long as it is manageable for you to 1) consolidate input/edits and 2) port to markdown

@RobertBroersma what I feel like I'm missing is an outline with a list of topics/examples to cover. One way this will become apparent is when we update the current generator test.js templates. But another way to get there would be to schedule a time for you to walk us through new testing capabilities in detail. (Which you might have already done with Peter.)

The test work + support is really fantastic and it's going to be a huge lift for the Redwood community. It's definitely worth the effort to prioritize comprehensive documentation.

That said, what's the most effective next step to get on the same page? And _when_ is a good time for the next step.

**Comment by RobertBroersma** [comment]
@thedavidprice I definitely think comprehensive testing guidelines are important for the Redwood ecosystem. Writing tests can be quite daunting for people who are new to testing, even without Cells, MockedProviders and whatnot, so I think some nice _opinionated_ guidelines could be very helpful. Some opinions will also implicitly be worked into the generated templates of course!

I think it's good for me to take a step back from the code and think about what a new user would actually want/need to know about testing to make good use out of this tooling.

Let's schedule a time for the walkthrough indeed and I think it could be helpful to draw some inspiration from other testing docs, like [Gatsby's](https://www.gatsbyjs.org/docs/testing/)!

**Comment by guledali** [comment]
@RobertBolender How would you go about testing protected routes?

**Comment by RobertBroersma** [comment]
@guledali That's a good point actually. I haven't looked into those yet, but I suspect the testing package needs some minor additions for those to function properly.

After that you can test them as any other route :)

**Comment by guledali** [comment]
BTW @RobertBroersma  Good work on testing 👍 Can't wait for the API side

**Comment by jtoar** [comment]
@RobertBroersma Circling back to this today. 

> Google docs process seems fine as long as it is manageable for you to 1) consolidate input/edits and 2) port to markdown

@thedavidprice Yes to both of these. 

> What I feel like I'm missing is an outline with a list of topics/examples to cover.

Definitely feeling this. But I'm also not very familiar with the work that's been going on. I'm going to take the time today to remedy that.

If that walkthrough hasn't happened yet, I'm up for scheduling that as well.

**Comment by RobertBroersma** [comment]
Bawnjorno,

This here PR https://github.com/redwoodjs/redwood/pull/805 introduced some more testing stuff that could use some docs!

## Running Tests

It now works similar to `create-react-app`:
- Running `yarn rw test` in your project automatically runs tests related to changed files.
- Running `yarn rw test --watchAll` automatically runs ALL tests on changing ANY file.
- Running `yarn rw test --coverage` will run all tests once, collect coverage and generate a report for your entire app, all sides.
- Running `yarn rw test` in CI will run tests only once. (Simulate by running `CI=1 yarn rw test`)
- You can run any of the above commands with one or more `--side <folderWhereSideResides>` to isolate tests to particular sides.

## Extending (Jest) Configs
You can now extend our default Jest configs using a generic `getConfig` method. In fact, importing our base configs is required. Example:
```js
const { getConfig } = require('@redwoodjs/core')

const config = getConfig({ type: 'jest', target: 'browser' })

// Optionally set the name of the side. This might be done automatically in the future.
config.displayName.name = 'web'

module.exports = config
```

**Comment by thedavidprice** [comment]
Thanks for this @RobertBroersma!

Are there instructions or a specific example about the Jest config files+code that will need to be added to a Redwood App? I'll need to include them in the next release.

**Comment by RobertBroersma** [comment]
@thedavidprice If we end up *not* having the upgrade codemod, I think (almost) the exact thing in my comment under *Extending (Jest) configs should be included. Maybe slightly different wording.
